decode: define BERT module instead of trying to look it up

carlosmn opened the pull request:
We want to make sure we are dealing with a module as we expect. Looking up the
object seems to result in the runtime re-using the returned `VALUE` leading to
us calling `supports?` on arbitrary objects when trying to decode.

The existing pattern for re-opening modules is to use `rb_define_module` so
let's do this.

/cc @github/git-systems for gitrpc
/cc @github/ruby-architecture as experts at ruby

---

We found this in gitauth which are pods which can spend some time without trying to do GitRPC and thus BERT which might explain why this could fall out of being loaded, perhaps. The error messages are like

```
message="NoMethodError: undefined method `supports?' for \\"va3-iad\\":String"
message="NoMethodError: undefined method `supports?' for []:Array"
message="NoMethodError: undefined method `supports?' for \\"parent_span_guid\\":String"
```

pointing to the code which calls `Decode#decode` which is what made me look for random memory access.

I don't know if the `rb_const_get` we have to get `BERT::Tuple` is safe to retain or not but this should already be better.

carlosmn committed:
decode: define BERT module instead of trying to look it up

We want to make sure we are dealing with a module as we expect. Looking up the
object seems to result in the runtime re-using the returned `VALUE` leading to
us calling `supports?` on arbitrary objects when trying to decode.

The existing pattern for re-opening modules is to use `rb_define_module` so
let's do this.

carlosmn commented:
The code trying to call `supports?` is https://github.com/github/bert/blob/f6727fc37b81fc278cb28d4b99c7e931bebe3609/ext/bert/c/decode.c#L516-L519

tgummerer reviewed (approved):
I don't know enough ruby to say if this is the right fix, but the explanation is plausible. Thanks for digging into this

carlosmn commented:
Looking through the rugged code, which I trust to work reasonably well, the only time `rb_const_get` is used is to get a module and it is used within a single block, so I think more exploration is needed in order to see if we really can keep `rb_cTuple` around.

carlosmn commented:
Based on looking around code, including some shipped with ruby itself, I believe we should not be keeping the results of `rb_const_get` past a function call. See e.g. https://github.com/ruby/ruby/blob/48b94b791997881929c739c64f95ac30f3fd0bb9/ext/openssl/ossl_x509name.c#L27-L30 where a pair of macros are defined to be used in different parts of the file instead of storing the pointer (those macros have been there for 17 years so I'm going to trust that that's the right thing to do).

carlosmn committed:
decode: do not keep a reference to classes across calls

We do not want to keep the `Bert::Tuple` and `Mochilo` references and re-use
those values across calls. This has never been something we should do, but with
the advent of the compacting GC this becomes even more important.

Look up these values if we need them to be used only inside a single call into
the C code.
